### ISSUE TITLE
Warning: RelayConnectionHandler: Unexpected after cursor `WyJuYXR1cmFsIiwxMF0=`, edges must be fetched from the end of the list (`WyJuYXR1cmFsIiwyMF0=`).

### ISSUE THREAD

**Issue by jazblueful** [relay-tools/relay-hooks#28]
I got another strange bug I need to squash before I can carry on with other tasks.

On page ```http://my-app/my-posts``` there is a clickable element with onClick handler:
```js
(event) => {
  event.preventDefault();
  history.push('/my-posts'); // same page as current value in browser address bar
}
```
On the same page there is a component with ```Load More``` button. When clicking on that button, it calls loadMore function returned by usePagination hook. It's almost identical to the usePagination example.

If you press the ```Load More``` button without doing anything else, it loads another set of results from the graphql server without problems. It also works if you navigate to another page, and then back (doesn't matter if using browser's back button, or using another history.push call, it still works). But for some reason it fails after trying to navigate to the same page, you're already at, for example you're on page ```http://my-app/my-posts``` and call history.push('/my-posts') and then press the ```Load More ``` no new items are rendered and I get the following warning:

```index.js:1 Warning: RelayConnectionHandler: Unexpected after cursor `WyJuYXR1cmFsIiwxMF0=`, edges must be fetched from the end of the list (`WyJuYXR1cmFsIiwyMF0=`).```

I see that the graphql request and response are identical in both cases - when it works fine, and when it fails.

Here is the relevant stacktrace:

```
console.<computed> | @ | index.js:1
-- | -- | --
  | r | @ | backend.js:1
  | printWarning | @ | warning.js:30
  | push../node_modules/fbjs/lib/warning.js.warning | @ | warning.js:51
  | update | @ | RelayConnectionHandler.js:127
  | (anonymous) | @ | RelayPublishQueue.js:216
  | _getSourceFromPayload | @ | RelayPublishQueue.js:212
  | (anonymous) | @ | RelayPublishQueue.js:242
  | _commitData | @ | RelayPublishQueue.js:238
  | run | @ | RelayPublishQueue.js:178
  | _processResponse | @ | RelayModernQueryExecutor.js:315
  | _handleNext | @ | RelayModernQueryExecutor.js:263
  | (anonymous) | @ | RelayModernQueryExecutor.js:207
  | _schedule | @ | RelayModernQueryExecutor.js:178
  | _next | @ | RelayModernQueryExecutor.js:206
  | next | @ | RelayModernQueryExecutor.js:94
  | next | @ | RelayObservable.js:565
  | (anonymous) | @ | RelayNetworkLayer.js:55
  | Promise.then (async) |   |  
  | subscribe | @ | RelayNetworkLayer.js:54
  | (anonymous) | @ | RelayObservable.js:474
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | Executor | @ | RelayModernQueryExecutor.js:85
  | execute | @ | RelayModernQueryExecutor.js:43
  | (anonymous) | @ | RelayModernEnvironment.js:224
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | (anonymous) | @ | RelayObservable.js:233
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | (anonymous) | @ | fetchQueryInternal.js:114
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | (anonymous) | @ | RelayObservable.js:211
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | (anonymous) | @ | RelayObservable.js:369
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | (anonymous) | @ | RelayObservable.js:211
  | _subscribe | @ | RelayObservable.js:610
  | subscribe | @ | RelayObservable.js:296
  | push../node_modules/relay-hooks/lib/FragmentPagination.js.FragmentPagination._fetchPage | @ | FragmentPagination.js:346
  | push../node_modules/relay-hooks/lib/FragmentPagination.js.FragmentPagination.loadMore | @ | FragmentPagination.js:225
  | loadMore | @ | useOssFragment.js:94

```

Thanks for looking into it.

**Comment by sibelius** [comment]
can you share more code?

**Comment by jazblueful** [comment]
Hi sibelius, glad to see you still lurking around here :)

I'm using fetchQuery to imperatively fetch the query and all the fragments that are needed by the components rendered when navigating to that route. The result is then put into context which each component can then use ( I don't like passing props deep down to nested components). If you think it may be relevant, you can see the code here: https://github.com/relay-tools/relay-hooks/issues/27

Here is the usePagination part, which is pretty similar to the relay-hooks usePagination example:
```typescript
import { usePagination, graphql } from 'relay-hooks';
import { ConnectionConfig } from 'relay-hooks/lib/FragmentPagination';
import { PaginationFunction } from 'relay-hooks/lib/useOssFragment';

import { useRelayDataContext } from '@util';
import { usePrintsByUsernameQueryConnection } from './__generated__/usePrintsByUsernameQueryConnection.graphql';

export type Print = Exclude<usePrintsByUsernameQueryConnection['printsByUsernameConnection']['edges'][0]['node'], null>;

interface UseGetPrintsResult extends Pick<PaginationFunction, 'loadMore' | 'hasMore' | 'isLoading'> {
  connectionConfig: ConnectionConfig;
  prints: usePrintsByUsernameQueryConnection['printsByUsernameConnection'];
}

const fragmentSpec = graphql`
  fragment usePrintsByUsernameQueryConnection on Query
    @argumentDefinitions(
      username: { type: "String!" }
      count: { type: "Int", defaultValue: 10 }
      cursor: { type: "Cursor" }
    ) {
    printsByUsernameConnection(first: $count, after: $cursor, username: $username)
      @connection(key: "usePrintsByUsernameQueryConnection_printsByUsernameConnection") {
      edges {
        node {
          id
          appUserId
          alias
          isPublished
        }
      }
    }
  }
`;

const connectionConfig: ConnectionConfig = {
  getVariables(props, { count, cursor }, fragmentVariables) {
    return {
      username: fragmentVariables.username,
      count,
      cursor,
    };
  },
  query: graphql`
    query usePrintsByUsernameQuery($count: Int!, $cursor: Cursor, $username: String!) {
      ...usePrintsByUsernameQueryConnection @arguments(count: $count, cursor: $cursor, username: $username)
    }
  `,
};

export const usePrintsByUsernameQuery = function(): UseGetPrintsResult {
  const relayData = useRelayDataContext();
  const [{ printsByUsernameConnection }, { isLoading, hasMore, loadMore }] = usePagination(fragmentSpec, relayData);

  return {
    isLoading,
    hasMore,
    loadMore,
    connectionConfig,
    prints: printsByUsernameConnection,
  };
};

export default usePrintsByUsernameQuery;
```

clickHandler on the component which uses usePrintsByUsernameQuery hook:
```typescript
() => {
  if (!hasMore() || isLoading()) {
    return;
  }
  loadMore(
    connectionConfig,
    10,
    (error: Error) => {
      if (error) console.log(error);
    },
    {},
  );
}
```

Hope it helps

**Comment by sibelius** [comment]
are you passing `cursor` to `loadMore`?

you need to pass the endCursor to loadMore to make sure if will fetch from the end of the connection

check if this hook helps you https://gist.github.com/sibelius/22d2690dd1146a96ebeda47f9a664efd

> I'm using fetchQuery to imperatively fetch the query and all the fragments that are needed by the components rendered when navigating to that route

why are you doing this?

can you share this part of the code as well?

**Comment by jazblueful** [comment]
I'm using [universal-router](https://github.com/kriasoft/universal-router) and it's imperative, not declarative like react-router is. I want to ensure the query (and all the fragments from child components) are loaded before rendering the root route component for server-side-rendering to work.

**Comment by morrys** [comment]
In fact, it's not necessary, you used the cursor value of the getVariables function ...

In reality, the call's response is strange, if you notice inside pageInfo cursor it is still set to WyJuYXR1cmFsIiwyMF0=

**Comment by morrys** [comment]
This leads you to make another loadMore but in reality you are loading data from an old paging.

**Comment by jazblueful** [comment]
According to [relay connections specification](https://facebook.github.io/relay/graphql/connections.htm)

> PageInfo must contain fields hasPreviousPage and hasNextPage, both of which return non‐null booleans. **It must also contain fields startCursor and endCursor, both of which return non‐null opaque strings.**

and 

> startCursor and endCursor must be the cursors corresponding to the first and last nodes in edges, respectively

So the value of endCursor cannot be null

**Comment by morrys** [comment]
https://github.com/facebook/relay/pull/2655

**Comment by sibelius** [comment]
`hasNextPage` should be enough to check if we need to refetch more pages or not

**Comment by jazblueful** [comment]
@morrys their case is when edges return 0 nodes, endCursor is null. But in my case, there are nodes and the endCursor has a correct value of the cursor of the last node in edges.

**Comment by morrys** [comment]
sorry :) they seemed to me the same :D

WyJuYXR1cmFsIiwxMF0
WyJuYXR1cmFsIiwyMF0

**Comment by jazblueful** [comment]
It's easy mistake to make, only 1 letter difference :)

WyJuYXR1cmFsIiw_x_MF0
WyJuYXR1cmFsIiw_y_MF0

**Comment by jazblueful** [comment]
1 thing I forgot to mention is that I need to press ```Load More``` button twice for the warning to be printed on the console.

**Comment by morrys** [comment]
I try to respond to the warning: the first query is executed, updates the store but does not refresh the UI, the second is executed but recovered from the cache (network) and when it updates in the store the warning is returned because it finds the data from the first query.

The real question is why the UI in the first query is not updated.

What do you say?

**Comment by jazblueful** [comment]
You may be onto something here.

I'm comparing the RecordSource between:
- A: (left in the diff) after initial navigation to page with the ```Load More``` button  (without pressing the button)
- B: (right in the diff) after navigating to the same page (without pressing the button) 

Diff: http://www.mergely.com/qQDApk1A

I spotted something strange here: ```"__connection_next_edge_index"``` is 30 after navigating to the same page while there are only 20 records in the database and there are some other differences.

Graphql query is sent to the api only when initially navigating to the page, not when navigating to the same page again, which is correct behavior as the data should already be in RecordSource after initial navigation. Then from where does the difference in RecordSources come from? Shouldn't they be identical?

**Comment by morrys** [comment]
Have you re-run the query and haven't done the dispose of the one before?

**Comment by jazblueful** [comment]
yeah, navigating to the same page, reruns the same query:

```typescript
      relayData = (await fetchQuery(environment, route.query, variables /*, { force: true }*/)) as RelayData;
      // fetchQuery bypass the store and is stored in recordSource directly.
      // Garbage collector would dispose the data in recordSource,
      // unless we explicitly tell it to to retain it.
      // https://github.com/relay-tools/relay-hooks/issues/27
      const request = getRequest(route.query);
      const operation = createOperationDescriptor(request, variables);
      environment.retain(operation.root);
     // used output from the following console.log for diffing
      console.log(
        JSON.stringify(
          environment
            .getStore()
            .getSource()
            .toJSON(),
          undefined,
          2,
        ),
      );
```

**Comment by morrys** [comment]
Surely you should manage the dispose of the previous query correctly. So as to maintain a more consistent situation.

Then I can advise you to log what is rendered and what is published in the store (publish function of the store).

**Comment by jazblueful** [comment]
Clearly I don't have enough knowledge to do this on my own right now, can you point me to the relevant documentation/article/issue that would help me? Relay's documentation for fetchQuery isn't particularly informative.

**Comment by morrys** [comment]
`request = environment.retain(operation.root);`
when unmount  the component: `request.dispose()`

maybe @sibelius can better point you to some articles.

**Comment by sibelius** [comment]
check this

https://babangsund.com/relay_environment_methods/

and

https://babangsund.com/relay_environment_methods_2/

**Comment by jazblueful** [comment]
I'm now disposing the retained request before every subsequent request, but the result is exactly the same :(

Perhaps the issue is that dispose tells garbage collector "it's ok to remove data that resulted with this request", but doesn't guarantee garbage collector will actually run before the next request.

```typescript
import React from 'react';
import { createOperationDescriptor, fetchQuery, getRequest, Disposable } from 'relay-runtime';

import { RelayDataContext, RelayData, RouteDataContext } from '@util';
import { Route, RouteContext } from '@core/router';

let disposable: Disposable | undefined;

export const dataProviderMiddleware = async (route: Route, context: RouteContext) => {
  const { environment, history } = context;
  const routeData = {
    history,
    params: route.params,
    query: context.query,
  };
  const variables = route.variables || {};

  let relayData = null;

  if (route.query) {
    try {
      if (disposable) {
        disposable.dispose();
      }
      relayData = (await fetchQuery(environment, route.query, variables, { force: true })) as RelayData;
      // fetchQuery bypasses the store and is stored in RecordSource directly.
      // Garbage collector would dispose the data in RecordSource,
      // unless we explicitly tell it to to retain it.
      // https://github.com/relay-tools/relay-hooks/issues/27
      const request = getRequest(route.query);
      const operation = createOperationDescriptor(request, variables);
      disposable = environment.retain(operation.root);
    } catch (err) {
      console.error(err);
    }
  }

  return {
    component: (
      <RouteDataContext.Provider value={routeData}>
        <RelayDataContext.Provider value={relayData}>{route.component}</RelayDataContext.Provider>
      </RouteDataContext.Provider>
    ),
  };
};

export default dataProviderMiddleware;

```

**Comment by morrys** [comment]
I must admit that this issue is slightly complex :)

The fact that you see the index increase is not worrying, the relay store uses the ref and _id fields to reference nodes. And you can see that in the second file the `client:root:__usePrintsByUsernameQueryConnection_printsByUsernameConnection_connection(username:\"reanna\"):edges:0` up to `client:root:__usePrintsByUsernameQueryConnection_printsByUsernameConnection_connection(username:\"reanna\"):edges:19` are not most referenced by anyone.

This means that the garbage collector can eliminate them.

Your problem seems to be related to the fetchQuery and the renderer, but perhaps the state of the store is useful when the warning occurs and specify which node you are viewing last in the browser.

**Comment by jazblueful** [comment]
It's complex indeed :p

I don't think it's a renderer issue because usePagination doesn't trigger a renderer, even though I see a query being successfully sent to the server.

What do you think?

**Comment by morrys** [comment]
Could you summarize the operations you do when you detect the warning and when it works?

Moreover, you could debug in these two points in both cases.

https://github.com/relay-tools/relay-hooks/blob/master/src/FragmentPagination.ts#L471

https://github.com/relay-tools/relay-hooks/blob/master/src/useOssFragment.tsx#L102

**Comment by jazblueful** [comment]
There is a difference there:
https://github.com/relay-tools/relay-hooks/blob/80c7203ee21c4a8b9a186ef506812761628e4ba8/src/useOssFragment.tsx#L103

When ```Load More```` works fine, ```result.data``` contains first 10 items and ```newData``` contains first 10 items + 10 newly loaded items (total of 20) so the if check passes and it triggers a rerender.

When ```Load More``` doesn't work, both ```result.data``` and ```newData``` contain the same 10 items that were loaded before pressing the ```Load More``` button.

To summarize:

A: (```Load More```` works)
Initial rendering of the page as soon as fetchQuery has a result from the server.
List of 10 items appear with ```Load More``` button
Pressing the ```Load More``` button, a graphql request for next items is sent/received successfully
Another 10 items are now displayed (total of 20).
```Load More``` button is now gone as we have reached the end of the list.

B: (```Load More``` doesn't work)
Initial rendering of the page as soon as fetchQuery has a result from the server.
List of 10 items appear with ```Load More``` button
Press menu item that does history.push to the current page
Pressing the ```Load More``` button, a graphql request for next items is sent/received successfully
Nothing changes, nothing is printed to the console
Pressing the ```Load More``` button again, no request is sent, but now there is a warning in the console.

**Comment by jazblueful** [comment]
https://github.com/relay-tools/relay-hooks/blob/80c7203ee21c4a8b9a186ef506812761628e4ba8/src/FragmentPagination.ts#L471

For neither A or B, the if check doesn't pass in FragmentPagination. Both prevData and nextData contain 20 items, both for A and B.

**Comment by morrys** [comment]
It might be useful to make the diff between the store content in these two scenarios at [this point](https://github.com/relay-tools/relay-hooks/blob/80c7203ee21c4a8b9a186ef506812761628e4ba8/src/useOssFragment.tsx#L103)
```
When Load More works fine, result.data contains first 10 items and newData contains first 10 items + 10 newly loaded items (total of 20) so the if check passes and it triggers to rerender.
```
diff
```
When Load More doesn't work, both result.data and newData contain the same 10 items that were loaded before pressing the Load More button.
```

**Comment by jazblueful** [comment]
http://www.mergely.com/oGJrCpNq/
Left: working
Right: not working

**Comment by morrys** [comment]
The store "seems" correct ... Could you do the diff of the serialized object "resolver" here?
https://github.com/relay-tools/relay-hooks/blob/80c7203ee21c4a8b9a186ef506812761628e4ba8/src/useOssFragment.tsx#L102

**Comment by jazblueful** [comment]
http://www.mergely.com/W4x59MCT/
Left: working
Right: not working

**Comment by morrys** [comment]
You sent the diff of the data not of the resolver :D

**Comment by jazblueful** [comment]
Ah, sorry about that :D

http://www.mergely.com/TaK94kyn/
Left: working
Right: not working

**Comment by morrys** [comment]
The problem could be in this difference, now we need to understand how this is caused.
```
      "_subscription": null
    }
  },
  "_stale": false
}
```
Certainly adding a log in this class https://github.com/facebook/relay/blob/v5.0.0/packages/relay-runtime/store/RelayModernFragmentSpecResolver.js can help, especially where the variable stale is changed.

**Comment by morrys** [comment]
Can you even try to change this?
https://github.com/relay-tools/relay-hooks/blob/master/src/useOssFragment.tsx#L85-L87

in this way:
`const [result, setResult] = useState<ContainerResult>(newResolver());`

**Comment by jazblueful** [comment]
Finally we are starting to get somewhere :)

That change fixed it for when there are 20 items total (2 pages).

With 30 items, I get the same warning when trying to load the 3rd page.

**Comment by morrys** [comment]
The real problem is that I can't create your case in my project, it always seems to work for me .. there is something in your use case that escapes.

You can try adding
this `console.log ("callback");` here
https://github.com/relay-tools/relay-hooks/blob/master/src/useOssFragment.tsx#L102
and this `console.log ("callback onNext");` here
https://github.com/relay-tools/relay-hooks/blob/master/src/FragmentPagination.ts#L470


without the first change and  tell me if the number of invocations and the order is unchanged in the two cases?

**Comment by jazblueful** [comment]
I reverted the change I made earlier and changed total of items back to 20.

Working output:
```
callback
callback onNext
```
Not working output (after first load more button press):
```
callback
callback onNext
```
Not Working output (after second load more button press):
```
callback
callback onNext
Warning: RelayConnectionHandler: Unexpected after cursor `WyJuYXR1cmFsIiwxMF0=`, edges must be fetched from the end of the list (`WyJuYXR1cmFsIiwyMF0=`).
callback onNext
```

**Comment by morrys** [comment]
Only One?

**Comment by jazblueful** [comment]
Only one, yes.

**Comment by morrys** [comment]
The situation is too strange, it seems that old data is recovered in the callback and data updated in the onNext function, as if there were two different stores, and from the information you gave me it all seems to work properly.

At the moment I can only ask you to recreate the error in a project where I can try or you have to investigate the situation more, adding console.log in the classes that I have indicated to you, and trying to understand what changes when you push (tried to avoid the push if you are already on the current page?).

**Comment by jazblueful** [comment]
I wrote a unit test to reproduce the bug (should have done it way sooner):
```typescript
import { render, cleanup, act, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
// @ts-ignore
import { createMockEnvironment, MockPayloadGenerator } from 'relay-test-utils';

import React, { FC } from 'react';
import { createOperationDescriptor, fetchQuery, getRequest, graphql, OperationDescriptor } from 'relay-runtime';
import { RelayEnvironmentProvider } from 'relay-hooks';

import { RouteDataContext, RouteData, RelayData, RelayDataContext } from '@util';
import { UserPrints } from '@prints/containers/UserPrints';

const query = graphql`
  query UserPrintsTestQuery($username: String!) {
    ...usePrintsByUsernameQueryConnection @arguments(username: $username)
  }
`;

let disposable;

const setup = (routeData?: RouteData) => {
  const environment = createMockEnvironment();

  if (!routeData) {
    // @ts-ignore
    routeData = {
      params: {
        username: 'reanna',
      },
    };
  }

  return {
    routeData,
    environment,
  };
};

afterEach(() => {
  cleanup();
});

test('Should render the component', async function() {
  const { routeData, environment } = setup();

  const variables = { username: 'reanna' };

  let fetchPromise = fetchQuery(environment, query, variables /*, { force: true }*/);

  expect(environment.mock.getAllOperations().length).toBe(1);

  act(() => {
    environment.mock.resolveMostRecentOperation((operation: OperationDescriptor) =>
      MockPayloadGenerator.generate(operation, {
        ID(_, generateId) {
          // Why we're doing this?
          // To make sure that we will generate a different set of ID
          // for elements on first page and the second page.
          return `first-page-id-${generateId()}`;
        },
        PageInfo() {
          return {
            hasNextPage: true,
          };
        },
      }),
    );
  });

  let relayData = (await fetchPromise) as RelayData;
  // fetchQuery bypasses the store and is stored in RecordSource directly.
  // Garbage collector would dispose the data in RecordSource,
  // unless we explicitly tell it to to retain it.
  // https://github.com/relay-tools/relay-hooks/issues/27
  let request = getRequest(query);
  let operation = createOperationDescriptor(request, variables);
  disposable = environment.retain(operation.root);

  const Component: FC<{}> = () => (
    <RelayEnvironmentProvider environment={environment}>
      <RouteDataContext.Provider value={routeData as RouteData}>
        <RelayDataContext.Provider value={relayData}>
          <UserPrints />
        </RelayDataContext.Provider>
      </RouteDataContext.Provider>
    </RelayEnvironmentProvider>
  );

  const { rerender, getByText } = render(<Component />);

  const loadMoreButton = getByText(/load more/i);
  expect(loadMoreButton).toBeInTheDocument();

  disposable.dispose();
  fetchPromise = fetchQuery(environment, query, variables /*, { force: true }*/);

  act(() => {
    environment.mock.resolveMostRecentOperation((operation: OperationDescriptor) =>
      MockPayloadGenerator.generate(operation, {
        ID(_, generateId) {
          return `first-page-id-${generateId()}`;
        },
        PageInfo() {
          return {
            hasNextPage: true,
          };
        },
      }),
    );
  });

  relayData = (await fetchPromise) as RelayData;

  rerender(<Component />); // commenting this line out, will make the last assertion pass

  fireEvent.click(loadMoreButton);

  act(() => {
    environment.mock.resolveMostRecentOperation((operation: OperationDescriptor) =>
      MockPayloadGenerator.generate(operation, {
        ID(_, generateId) {
          return `second-page-id-${generateId()}`;
        },
        PageInfo() {
          return {
            hasNextPage: false,
          };
        },
      }),
    );
  });

  expect(loadMoreButton).not.toBeInTheDocument();
});
```

It looks it's caused by rerendering the component after refetching the same data. I will extract the test and all the relevant code into a separate repo tomorrow for you to play with.

**Comment by morrys** [comment]
yours is good news;) from tomorrow to monday i'm on vacation, i can watch it but don't try it ... I hope that we can resolve It today;)

**Comment by jazblueful** [comment]
Thank you so much for trying to get it resolved before you need to go.

Here is the repo as promised: https://github.com/jazblueful/loadmore-rerender

**Comment by morrys** [comment]
Found the problem :)
https://github.com/relay-tools/relay-hooks/blob/master/src/useOssFragment.tsx#L102
In the callback there is the reference to resolver and not to res. Then it used the old disposed resolver :)

I have to do more tests :)

But you can already try using this:
```
res.setCallback (function () {
             var newData = res.resolve();
             if (result.data! == newData) {
                 setResult ({resolver: res, data: newData});
             }
         });
```

**Comment by jazblueful** [comment]
Yep, this fixes it for me.

Thank you! :tada:

**Comment by morrys** [comment]
I was sure we would arrive at the solution: D your test case was fundamental (it is better that we do not say this to @sibelius ) :)

Now release version 1.2.6

I close the issue after you've tried the new version ok? :)

**Comment by morrys** [comment]
https://github.com/relay-tools/relay-hooks/releases/tag/v1.2.6

:D

**Comment by morrys** [comment]
https://github.com/relay-tools/relay-hooks/releases/tag/v1.2.7

The final test did not pass ... another change was needed.

https://github.com/relay-tools/relay-hooks/releases/tag/v1.2.7

Let me know if everything works, that I still have some time in the afternoon :)

**Comment by jazblueful** [comment]
I'm glad you found the test case helpful :)

Relay hooks is now working perfectly. Only problem is, the dispose functionality I introduced while trying to figure out the root cause of this issue, is breaking my app when navigating away from a route and then back to it, but that's unrelated to this issue.

Thank you very much once again and don't worry, @sibelius won't hear a thing from me :D

Have a great vacation and if you see my mother-in-law and my wife's sister, who are currently on vacation in Italy, say hi to them :)

**Comment by morrys** [comment]
Perfect, if you can keep track of the change made in version 1.2.7, if you have problems write as well :)
